File: sdks/typescript/examples/virtual-mic.ts
import { setTimeout as delay } from "node:timers/promises";

import { MarsClient } from "../src/index.js";

const sampleRate = 48_000;
const chunkFrames = 480;
const client = new MarsClient();
const outcome = await client.setVirtualInputs("com.example.typescript", [{
  id: "mic",
  name: "TypeScript Virtual Mic",
  uid: "com.example.typescript.mic",
  sampleRate,
  channels: 1,
}]);
const mic = outcome.virtualMics[0];
if (mic === undefined) throw new Error("daemon did not return a virtual mic handle");

const writer = mic.openLiveWriter();
const chunk = new Float32Array(chunkFrames);
let phase = 0;

try {
  for (let index = 0; index < 500; index += 1) {
    for (let frame = 0; frame < chunkFrames; frame += 1) {
      chunk[frame] = 0.25 * Math.sin(phase);
      phase = (phase + 2 * Math.PI * 440 / sampleRate) % (2 * Math.PI);
    }
    writer.writeF32InterleavedLive(chunk);
    await delay(10);
  }
  writer.flushSilence();
} finally {
  writer.close();
}

File: sdks/python/examples/virtual_mic.py
import asyncio
from array import array
import math

from mars_sdk import AppVirtualInputSpec, MarsClient


async def main() -> None:
    sample_rate = 48_000
    chunk_frames = 480
    client = MarsClient()
    outcome = await client.set_virtual_inputs(
        "com.example.python",
        [
            AppVirtualInputSpec(
                id="mic",
                name="Python Virtual Mic",
                uid="com.example.python.mic",
                sample_rate=sample_rate,
                channels=1,
            )
        ],
    )

    writer = outcome.virtual_mics[0].open_live_writer()
    chunk = array("f", [0.0]) * chunk_frames
    phase = 0.0
    try:
        for _ in range(500):
            for frame in range(chunk_frames):
                chunk[frame] = 0.25 * math.sin(phase)
                phase = (phase + math.tau * 440 / sample_rate) % math.tau
            writer.write_f32_interleaved_live(chunk)
            await asyncio.sleep(0.01)
        writer.flush_silence()
    finally:
        writer.close()


asyncio.run(main())
